packages/deployments/contracts/contracts/messaging/libraries/DomainIndexer.sol

```diff
@@ -0,0 +1,160 @@
+   * We should distribute the aggregate root to all of these domains in the `propagate` method.
+   * @dev Whenever this domains array is updated, the connectors array should also be updated.
+   */
+  uint32[] public domains;
+
+  /**
+   * @notice A "quick reference" hash used in the `propagate` method below to validate that the provided
+   * array of domains matches the one we have in storage.
+   * @dev This hash should be re-calculated whenever the domains array is updated.
+   */
+  bytes32 public domainsHash;
+
+  /**
+   * @notice Tracks the addresses of the hub connector contracts corresponding to subscribed spoke domains.
+   * The index of any given connector in this array should match the index of that connector's target spoke
+   * domain in the `domains` array above.
+   * @dev This should be updated whenever the domains array is updated.
+   */
+  address[] public connectors;
+
+  /**
+   * @notice A "quick reference" hash used in the `propagate` method below to validate that the provided
+   * array of connectors matches the one we have in storage.
+   * @dev This hash should be re-calculated whenever the connectors array is updated.
+   */
+  bytes32 public connectorsHash;
+
+  /**
+   * @notice Shortcut to reverse lookup the index by domain. We index starting at one so the zero value can
+   * be considered invalid (see fn: `isDomainSupported`).
+   * @dev This should be updated whenever the domains array is updated.
+   */
+  mapping(uint32 => uint256) private domainToIndexPlusOne;
+
+  // ============ Getters ============
+
+  /**
+   * @notice Convenience shortcut for supported domains. Used to sanity check adding new domains.
+   * @param _domain Domain to check.
+   */
+  function isDomainSupported(uint32 _domain) public view returns (bool) {
+    return domainToIndexPlusOne[_domain] != 0;
+  }
+
+  /**
+   * @notice Gets the index of a given domain in the domains and connectors arrays.
+   * @dev Reverts if domain is not supported.
+   * @param _domain The domain for which to get the index value.
+   */
+  function getDomainIndex(uint32 _domain) public view returns (uint256) {
+    uint256 index = domainToIndexPlusOne[_domain];
+    require(index != 0, "!supported");
+    return index - 1;
+  }
+
+  /**
+   * @notice Gets the corresponding hub connector address for a given spoke domain.
+   * @dev Inefficient, should only be used by caller if they have no index reference.
+   * @param _domain The domain for which to get the hub connector address.
+   */
+  function getConnectorForDomain(uint32 _domain) public view returns (address) {
+    return connectors[getDomainIndex(_domain)];
+  }
+
+  /**
+   * @notice Validate given domains and connectors arrays are correct (i.e. they mirror what is
+   * currently saved in storage).
+   * @dev Reverts if domains or connectors do not match.
+   */
+  function validateDomains(uint32[] calldata _domains, address[] calldata _connectors) public view {
```

Comment:
just note that it will error if the ordering of the domains / connectors as well